dot/sync/worker_test.go

```diff
@@ -32,30 +40,47 @@ func TestWorker(t *testing.T) {
 	w := newWorker(peerA, sharedGuard, reqMaker)
 
 	wg := sync.WaitGroup{}
-	queue := make(chan *syncTask, 2)
+	queue := make(chan *syncTask, 1) // 1 means that maximum 1 worker can send request at a given time
 
-	wg.Add(1)
+	// run two workers, but they shouldn't work concurrently,
+	// because sharedGuard is buffered channel with capacity
+	wg.Add(2)
+	go w.run(queue, &wg)
 	go w.run(queue, &wg)
```

Comment:
There is a problem here, you're not runnning two workers but using the same worker instance to spawn another `run` method.

I would suggest to:

```
peerA := peer.ID("peerA")
peerB := peer.ID("peerB")

reqMaker.EXPECT().
		Do(peerA, nil, gomock.AssignableToTypeOf((*network.BlockResponseMessage)(nil))).
		DoAndReturn(func(_, _, _ any) any {
			time.Sleep(4 * time.Second)
			return nil
		}).
		Return(nil)

// define a mock expectation to peerB
reqMaker.EXPECT().
		Do(peerB, nil, gomock.AssignableToTypeOf((*network.BlockResponseMessage)(nil))).
		DoAndReturn(func(_, _, _ any) any {
			time.Sleep(4 * time.Second)
			return nil
		}).
		Return(nil)
		
// instantiate the workers
fstWorker := newWorker(peerA, sharedGuard, reqMaker)
sndWorker := newWorker(peerB, sharedGuard, reqMaker)

// and then run both
wg.Add(2)
go fstWorker.run(queue, &wg)
go sndWorker.run(queue, &wg)
```

Reply:
yes, that is my mistake. I will fix that